services/motion/builtin/moveonglobe.go

```diff
@@ -0,0 +1,314 @@
+	"sync"
+	"time"
+
+	geo "github.com/kellydunn/golang-geo"
+	"github.com/pkg/errors"
+	goutils "go.viam.com/utils"
+
+	"go.viam.com/rdk/components/base"
+	"go.viam.com/rdk/components/base/kinematicbase"
+	"go.viam.com/rdk/components/movementsensor"
+	"go.viam.com/rdk/motionplan"
+	"go.viam.com/rdk/referenceframe"
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/services/motion"
+	"go.viam.com/rdk/spatialmath"
+)
+
+type replanResponse struct {
+	err    error
+	replan bool
+}
+
+type executeResponse struct {
+	err     error
+	success bool
+}
+
+type planSession struct {
+	plan              motionplan.Plan
+	executionChan     chan executeResponse
+	positionChan      chan replanResponse
+	obsticleChan      chan replanResponse
+	cancelCtx         context.Context
+	cancelFn          context.CancelFunc
+	backgroundWorkers *sync.WaitGroup
+}
+
+func plan(
+	ctx context.Context,
+	planRequest *motionplan.PlanRequest,
+	kb kinematicbase.KinematicBase,
+	componentName string,
+) (motionplan.Plan, error) {
+	inputs, err := kb.CurrentInputs(ctx)
+	if err != nil {
+		return make(motionplan.Plan, 0), err
+	}
+	// TODO: this is really hacky and we should figure out a better place to store this information
+	if len(kb.Kinematics().DoF()) == 2 {
+		inputs = inputs[:2]
+	}
+	planRequest.StartConfiguration = map[string][]referenceframe.Input{componentName: inputs}
+
+	return motionplan.PlanMotion(ctx, planRequest)
+}
+
+func startPollingForReplan(
+	ctx context.Context,
+	period time.Duration,
+	responseChan chan replanResponse,
+	backgroundWorkers *sync.WaitGroup,
+	fn func(context.Context,
+	) replanResponse,
+) {
+	backgroundWorkers.Add(1)
+	goutils.ManagedGo(func() {
+		ticker := time.NewTicker(period)
+		defer ticker.Stop()
+		for {
+			// this ensures that if the context is cancelled we always
+			// return early at the top of the loop
+			if err := ctx.Err(); err != nil {
+				return
+			}
+
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				response := fn(ctx)
+				if response.err != nil || response.replan {
+					responseChan <- response
+					return
+				}
+			}
+		}
+	}, backgroundWorkers.Done)
+}
+
+func (ps *planSession) start(
+	motionCfg *motion.MotionConfiguration,
+	kb kinematicbase.KinematicBase,
+	ms *builtIn,
+	movementSensor movementsensor.MovementSensor,
+	destination *geo.Point,
+	replanCount int,
+) {
+	positionPollingPeriod := time.Duration(1000/motionCfg.PositionPollingFreqHz) * time.Millisecond
+	obstaclePollingPeriod := time.Duration(1000/motionCfg.ObstaclePollingFreqHz) * time.Millisecond
+
+	// helper function to manage polling functions
+	// spawn two goroutines that each have the ability to trigger a replan
+	if positionPollingPeriod > 0 {
+		startPollingForReplan(
+			ps.cancelCtx,
+			positionPollingPeriod,
+			ps.positionChan,
+			ps.backgroundWorkers,
+			func(ctx context.Context) replanResponse {
+				if replanCount == 2 {
+					return replanResponse{replan: true}
+				}
+				return replanResponse{}
+			})
```

Comment:
This allows us to test the behavior of each polling go routine under test. We could also use this temporary replan count to test the behavior of the replan goroutines returning errors.